dpdata/abacus/scf.py

```diff
@@ -112,6 +112,98 @@ def get_cell(geometry_inlines):
+    - `angle2`: in non-colinear case, specify angle between a-axis and real spin in projection in ab-plane , in angle measure instead of radian measure
+
+      e.g.:
+
+      ```
+      Fe
+      1.0
+      2
+      0.0 0.0 0.0 m 0 0 0 mag 1.0 angle1 90 angle2 0 cs 0 0 0
+      0.5 0.5 0.5 m 1 1 1 mag 1.0 angle1 90 angle2 180
+    - `cs` or `constrain`: three numbers, which take value in 0 or 1, control the spin constraint of the atom.
+    - `lambda`: three numbers, control the lambda of the atom.
+    """
+    sline = pos_line.split()
+    pos = [float(i) for i in sline[:3]]
+    move = None
+    velocity = None
+    magmom = None
+    angle1 = None
+    angle2 = None
+    constrain = None
+    lambda1 = None
+    if len(sline) > 3:
+        mag_list = []
+        velocity_list = []
+        move_list = []
+        angle1_list = []
+        angle2_list = []
+        constrain_list = []
+        lambda_list = []
+        label = "move"
+        for i in range(3, len(sline)):
+            # firstly read the label
+            if sline[i] == "m":
+                label = "move"
+            elif sline[i] in ["v", "vel", "velocity"]:
+                label = "velocity"
+            elif sline[i] in ["mag", "magmom"]:
+                label = "magmom"
+            elif sline[i] == "angle1":
+                label = "angle1"
+            elif sline[i] == "angle2":
+                label = "angle2"
+            elif sline[i] in ["constrain", "sc"]:
+                label = "constrain"
+            elif sline[i] in ["lambda"]:
+                label = "lambda"
+
+            # the read the value to the list
+            elif label == "move":
+                move_list.append(int(sline[i]))
+            elif label == "velocity":
+                velocity_list.append(float(sline[i]))
+            elif label == "magmom":
+                mag_list.append(float(sline[i]))
+            elif label == "angle1":
+                angle1_list.append(float(sline[i]))
+            elif label == "angle2":
+                angle2_list.append(float(sline[i]))
+            elif label == "constrain":
+                constrain_list.append(bool(int(sline[i])))
+            elif label == "lambda":
+                lambda_list.append(float(sline[i]))
+        if len(move_list) == 3:
+            move = move_list
+        if len(velocity_list) == 3:
+            velocity = velocity_list
+        if len(mag_list) in [1, 3]:
+            magmom = mag_list if len(mag_list) == 3 else mag_list[0]
+        if len(angle1_list) == 1:
+            angle1 = angle1_list[0]
+        if len(angle2_list) == 1:
+            angle2 = angle2_list[0]
+        if len(constrain_list) == 3:
+            constrain = constrain_list
+        elif len(constrain_list) == 1:
+            constrain = constrain_list[0]
+        if len(lambda_list) == 3:
+            lambda1 = lambda_list
+        elif len(lambda_list) == 1:
+            lambda1 = lambda_list[0]
```

Comment:
This line is not covered by UT. 

error should be raised if no condition is satisfied.